File: 1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        if (root == null) return null;
        
        Queue<TreeNode[]> queue = new LinkedList<>();
        Map<Integer, TreeNode> nodeMap = new HashMap<>();
        Map<Integer, Integer> parentMap = new HashMap<>();

        queue.add(new TreeNode[] { root, null });
        List<List<Integer>> levels = new ArrayList<>();
        
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> level = new ArrayList<>();
            
            for (int i = 0; i < size; i++) {
                TreeNode[] pair = queue.poll();
                TreeNode node = pair[0], parent = pair[1];
                
                int x = node.val;
                int y = (parent == null) ? -1 : parent.val;
                
                level.add(x);
                parentMap.put(x, y);
                nodeMap.put(x, node);
                
                if (node.left != null) queue.add(new TreeNode[] { node.left, node });
                if (node.right != null) queue.add(new TreeNode[] { node.right, node });
            }

            levels.add(level);
        }

        List<Integer> lastLevel = levels.get(levels.size() - 1);
        if (lastLevel.size() == 1) return nodeMap.get(lastLevel.get(0));

        int s = lastLevel.get(0), t = lastLevel.get(lastLevel.size() - 1);
        while (s != t) {
            s = parentMap.get(s);
            t = parentMap.get(t);
        }

        return nodeMap.get(s);
    }
}

File: 1218-lowest-common-ancestor-of-deepest-leaves/lowest-common-ancestor-of-deepest-leaves.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func lcaDeepestLeaves(root *TreeNode) *TreeNode {
    if root == nil {
		return nil;
	}

	type Pair struct {
		Node *TreeNode;
		Parent *TreeNode;
	}

	queue := list.New();
	queue.PushBack(Pair{root, nil});

    parentMap := make(map[int]int);
    nodeMap := make(map[int]*TreeNode);

    var levels [][]int;

	for queue.Len() > 0 {
		size := queue.Len();
		level := make([]int, 0, size);

		for i := 0; i < size; i++ {
			front := queue.Remove(queue.Front()).(Pair);
			node := front.Node;
			parent := front.Parent;

			val := node.Val;
			parentVal := -1;
			if parent != nil {
				parentVal = parent.Val;
			}

			level = append(level, val);
			parentMap[val] = parentVal;
			nodeMap[val] = node;

			if node.Left != nil {
				queue.PushBack(Pair{node.Left, node});
			}

			if node.Right != nil {
				queue.PushBack(Pair{node.Right, node});
			}
		}

		levels = append(levels, level);
	}

    lastLevel := levels[len(levels)-1];
	if len(lastLevel) == 1 {
		return nodeMap[lastLevel[0]];
	}

	s, t := lastLevel[0], lastLevel[len(lastLevel)-1];

	for s != t {
		s = parentMap[s];
		t = parentMap[t];
	}

	return nodeMap[s];
}